File: cpp/04_005.cpp
#include <string>

class User; // Forward declaration
 
class UserRetrievalService {
public:
    virtual ~UserRetrievalService() = default;
    virtual User getUser(const std::string& username) = 0;
};
 
class UserManagementService {
public:
    virtual ~UserManagementService() = default;
    virtual void createUser(const User& user) = 0;
    virtual void updateUser(const User& user) = 0;
    virtual void deleteUser(const std::string& username) = 0;
};

File: data structures and algorithms/python/04_005.py
from abc import ABC, abstractmethod

class UserRetrievalService(ABC):
    @abstractmethod
    def get_user(self, username):
        pass
 
class UserManagementService(ABC):
    @abstractmethod
    def create_user(self, user):
        pass
    
    @abstractmethod
    def update_user(self, user):
        pass
    
    @abstractmethod
    def delete_user(self, username):
        pass